airflow/providers/google/cloud/triggers/dataproc.py

```diff
@@ -140,24 +142,97 @@ def serialize(self) -> tuple[str, dict[str, Any]]:
+        return await self.get_async_hook().get_cluster(
+            project_id=self.project_id, region=self.region, cluster_name=self.cluster_name
+        )
+
+    def check_cluster_state(self, state: ClusterStatus.State) -> bool:
+        """
+        Check if the state is error or running.
+
+        :param state: The state of the cluster.
+        """
+        return state in (ClusterStatus.State.ERROR, ClusterStatus.State.RUNNING)
+
+    async def gather_diagnostics_and_maybe_delete(self, cluster: Cluster):
+        """
+        Gather diagnostics and maybe delete the cluster.
+
+        :param cluster: The cluster to gather diagnostics for.
+        """
+        self.log.info("Cluster is in ERROR state. Gathering diagnostic information.")
+        try:
+            operation = await self.get_async_hook().diagnose_cluster(
+                region=self.region, cluster_name=self.cluster_name, project_id=self.project_id
             )
-            state = cluster.status.state
-            self.log.info("Dataproc cluster: %s is in state: %s", self.cluster_name, state)
-            if state in (
-                ClusterStatus.State.ERROR,
-                ClusterStatus.State.RUNNING,
-            ):
-                break
-            self.log.info("Sleeping for %s seconds.", self.polling_interval_seconds)
-            await asyncio.sleep(self.polling_interval_seconds)
-        yield TriggerEvent({"cluster_name": self.cluster_name, "cluster_state": state, "cluster": cluster})
+            result = await operation.result()
+            gcs_uri = str(result.response.value)
+            self.log.info(
+                "Diagnostic information for cluster %s available at: %s", self.cluster_name, gcs_uri
+            )
+        except Exception as e:
+            self.log.error("Failed to diagnose cluster: %s", e)
+
+        if self.delete_on_error:
+            await self.get_async_hook().delete_cluster(
+                region=self.region, cluster_name=self.cluster_name, project_id=self.project_id
+            )
+            return TriggerEvent(
+                {
+                    "cluster_name": self.cluster_name,
+                    "cluster_state": cluster.status.state,
+                    "cluster": None,
+                    "action": "deleted",
+                }
+            )
+        else:
+            return TriggerEvent(
+                {"cluster_name": self.cluster_name, "cluster_state": cluster.status.state, "cluster": cluster}
+            )
+
+    async def handle_cancellation(self) -> None:
+        """Handle the cancellation of the trigger, cleaning up resources if necessary."""
+        self.log.info("Cancellation requested. Deleting the cluster if created.")
+        try:
+            if self.delete_on_error:
+                cluster = await self.fetch_cluster_status()
+                if cluster.status.state == ClusterStatus.State.ERROR:
+                    await self.get_async_hook().async_delete_cluster(
```

Comment:
we call here `async_delete_cluster` on the hook, but in `gather_diagnostics_and_maybe_delete` we call a synchronous `delete_cluster`?

Reply:
I think this was a typo. I should be `delete_cluster` from async hook like in `gather_diagnostics_and_maybe_delete`